### DISCUSSION TITLE
PendingRollbackError in nested transactions

### DISCUSSION THREAD

**Discussion by csirmazbendeguz** [sqlalchemy/sqlalchemy#11071] [Usage Questions] [answered]
Hi!

I'm trying to test an `IntegrityError`. I've set up my tests to work with savepoints as per the [docs](https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#joining-a-session-into-an-external-transaction-such-as-for-test-suites).

```py
def test_user_model__unique_email(session: Session) -> None:
    user = UserFactory(email="rohan.kirby@example.com")

    with pytest.raises(IntegrityError) as exc_info:
        UserFactory(email="rohan.kirby@example.com")

    assert isinstance(exc_info.value.orig, psycopg2.errors.UniqueViolation)
    assert session.query(User).all() == [user]
"""
sqlalchemy.exc.PendingRollbackError: This Session's transaction has been rolled back due to a previous exception during flush. To begin a new transaction with this Session, first issue Session.rollback(). Original exception was: ...
"""
```

As you can see, an `IntegrityError` is raised on purpose during flush. Here's the SQL output:

```sql
BEGIN (implicit)
SAVEPOINT sa_savepoint_1
INSERT INTO users ...
RELEASE SAVEPOINT sa_savepoint_1
SAVEPOINT sa_savepoint_2
INSERT INTO users ...
ROLLBACK TO SAVEPOINT sa_savepoint_2
```

At this point, I would like to query the users again, to test a new user wasn't actually added (last line), but it raises a `PendingRollbackError`. I understand, since the nested transaction was rolled back automatically during the flush, the session state became `DEACTIVE`. Since I'm running my tests in transactions, I can't do a `ROLLBACK`.

I wonder how should I handle this?

**Comment by zzzeek** [comment, marked as answer]
this is hard to follow. is that actual code?  why would this raise an exception?

```
    with pytest.raises(IntegrityError) as exc_info:
        UserFactory(email="rohan.kirby@example.com")
```

i dont see any interaction with a Session there so it's not clear what you are actually doing.

Anytime a flush fails with the Session, [there must be an explicit call to rollback, if not using a context manager which does this](https://docs.sqlalchemy.org/en/14/faq/sessions.html#this-session-s-transaction-has-been-rolled-back-due-to-a-previous-exception-during-flush-or-similar), that is, whether or not you are inside of a test harness, code that throws an exception has to have a rollback to continue onwards.

What the recipe you reference does is turn the begin/commit/rollback sequence into one that is using a SAVEPPOINT, so that there is an outer BEGIN/ROLLBACK sequence that can't easily be escaped from, which is to allow for clean suite teardowns.

**Comment by csirmazbendeguz** [comment]
Thanks for the quick reply @zzzeek!

`UserFactory` is a `SQLAlchemyModelFactory` from [factory_boy](https://factoryboy.readthedocs.io/en/stable/orms.html#sqlalchemy), configured with `sqlalchemy_session_persistence = "commit"`, and `sqlalchemy_session` configured to be the same session as the one in the test, so that line is equal to a session add + session commit operation (autoflush enabled).

Actually, the issue was that I was not following the documentation properly, and so when I tried to explicitly call rollback in my test after the integrity error, it resulted in an outer `ROLLBACK`.

I didn't entirely understand how this worked:

```py
self.nested = self.connection.begin_nested()

@event.listens_for(self.session, "after_transaction_end")
def end_savepoint(session, transaction):
  if not self.nested.is_active:
    self.nested = self.connection.begin_nested()
```

So I replaced it with something that made more sense to me, but apparently it's incorrect 😅 :

```py
session.begin_nested()

@event.listens_for(session, "after_transaction_end")
def after_transaction_end(session: Session, transaction: SessionTransaction):
  if session.is_active:
    session.begin_nested()
```

Cheers!

**Reply by zzzeek** [reply]
> Thanks for the quick reply @zzzeek!
> 
> `UserFactory` is a `SQLAlchemyModelFactory` from [factory_boy](https://factoryboy.readthedocs.io/en/stable/orms.html#sqlalchemy), configured with `sqlalchemy_session_persistence = "commit"`, and `sqlalchemy_session` configured to be the same session as the one in the test, so that line is equal to a session add + session commit operation (autoflush enabled).

Oh OK, so first off I have no idea how that works, so not clear if it's interacting with the session in such a way that the recipe at https://docs.sqlalchemy.org/en/14/orm/session_transaction.html#joining-a-session-into-an-external-transaction-such-as-for-test-suites is not functioning correctly.  the recipe has two variants, the "plain" one that does not support rollback, and the savepoint version.   the recipe overall is [much improved, better and more predictable in SQLAlchemy 2.0](https://docs.sqlalchemy.org/en/20/orm/session_transaction.html#joining-a-session-into-an-external-transaction-such-as-for-test-suites) as well.

you'd want to get the recipe to work without using any third party extensions first; make sure you can do things with a `Session`, get integrity errors, call `session.rollback`, and after all that, the transaction that's on the `connection` from the outside is still in place.

> 
> Actually, the issue was that I was not following the documentation properly, and so when I tried to explicitly call rollback in my test after the integrity error, it resulted in an outer `ROLLBACK`.
> 
> I didn't entirely understand how this worked:
> 
> ```python
> self.nested = self.connection.begin_nested()
> 
> @event.listens_for(self.session, "after_transaction_end")
> def end_savepoint(session, transaction):
>   if not self.nested.is_active:
>     self.nested = self.connection.begin_nested()
> ```
> 
> So I replaced it with something that made more sense to me, but apparently it's incorrect 😅 :
> 
> ```python
> session.begin_nested()
> 
> @event.listens_for(session, "after_transaction_end")
> def after_transaction_end(session: Session, transaction: SessionTransaction):
>   if session.is_active:
>     session.begin_nested()
> ```

yeah if you can try out 2.0 where the behavior here that is quite confusing has been [made explicit](https://docs.sqlalchemy.org/en/20/orm/session_api.html#sqlalchemy.orm.Session.params.join_transaction_mode).

> 
> Cheers!

**Reply by csirmazbendeguz** [reply]
That's amazing, I was using 2.0 already but I didn't know it had a better way to do this. Thanks for all the help!

**Comment by csirmazbendeguz** [comment]
@zzzeek, may I have one more question related to this?

I'm trying to test a code with a transaction, and I'm receiving a `sqlalchemy.exc.InvalidRequestError: A transaction is already begun on this Session.` error.

Here's an example:
```
def test_db__transaction(session: Session) -> None:
    user = User(email="alasdai.haynes@example.com", password="abc")
    with session.begin():
        session.add(user)
```

As I understand, since the test is running in a transaction (as per above), new transactions can't be created. How can we work around this?

I'm trying to achieve something similar to Django's `TransactionTestCase`, which allowed testing this type of code (I'm not sure how it works though).

**Reply by zzzeek** [reply]
per the recipe in the docs, the `Session` is not "in a transaction" yet if this is working correctly.  `session.begin()` should succeed.  the session is bound to a connection, and that connection is in a transaction, but that's not the session's "transaction".  the `Session` has it's own "logical" transactional unit called `SessionTransaction` that does not imply any database connectivity.

would need to see a runnable example to understand why you'd get that error in this case.

**Reply by csirmazbendeguz** [reply]
I see, yes, you're right there is a select statement that runs first:

```
assert session.execute(select(func.count()).select_from(User)).scalar_one() == 0
```

This causes an `BEGIN (implicit)`, it start's the session transaction. If I remove this line, the test passes... but I would like to keep the line and it feels weird I need to roll back after a SELECT

**Reply by zzzeek** [reply]
why not move session.begin() above it?  That's the point of how the API is now.  we want you to know where your transaction starts.  if you have code like this:

```py
def run_my_thing():
    result = session.execute(stmt)
    # do other things, perhaps something time consuming
    with session.begin():
         # etc.
```

you're doing code review.  On what line of code did your real database transaction start?   If you want to make sure you dont have time consuming things in your transactions, how apparent above is it where the transaction actually is?    by having session.begin() just raise there, you will never have code like that above.  that's the idea!

**Reply by csirmazbendeguz** [reply]
I see, yeah, I think I'm getting the hang of it. Thanks for the explanation! I should always make sure to commit/rollback the session before I call the application code then (I'm trying to test an endpoint). Yeah, I was not expecting the first line to start the transaction.

**Reply by zzzeek** [reply]
right so unless you are using driver-level autocommit, all DBAPI statements implicitly start a transaction
